server/common/app_config.py

```diff
@@ -423,13 +574,114 @@ def handle_multi_dataset(self, context):
                 timelimit_s=self.multi_dataset__matrix_cache__timelimit_s,
             )
 
+    def handle_diffexp(self, context):
+        self.check_attr("diffexp__alg_cxg__max_workers", (str, int))
+        self.check_attr("diffexp__alg_cxg__cpu_multiplier", int)
+        self.check_attr("diffexp__alg_cxg__target_workunit", int)
+
+        max_workers = self.diffexp__alg_cxg__max_workers
+        cpu_multiplier = self.diffexp__alg_cxg__cpu_multiplier
+        cpu_count = os.cpu_count()
+        max_workers = min(max_workers, cpu_multiplier * cpu_count)
+        diffexp_tiledb.set_config(max_workers, self.diffexp__alg_cxg__target_workunit)
+
+    def handle_adaptor(self, context):
+        # cxg
+        self.check_attr("adaptor__cxg_adaptor__tiledb_ctx", dict)
+        regionkey = "vfs.s3.region"
+        if regionkey not in self.adaptor__cxg_adaptor__tiledb_ctx:
+            if type(self.data_locator__s3__region_name) == str:
+                self.adaptor__cxg_adaptor__tiledb_ctx[regionkey] = self.data_locator__s3__region_name
+
+        from server.data_cxg.cxg_adaptor import CxgAdaptor
+
+        CxgAdaptor.set_tiledb_context(self.adaptor__cxg_adaptor__tiledb_ctx)
+
+        # anndata
+        self.check_attr("adaptor__anndata_adaptor__backed", bool)
+
+    def handle_limits(self, context):
+        self.check_attr("limits__diffexp_cellcount_max", (type(None), int))
+        self.check_attr("limits__column_request_max", (type(None), int))
+
+    def exceeds_limit(self, limit_name, value):
+        limit_value = getattr(self, "limits__" + limit_name, None)
+        if limit_value is None:  # disabled
+            return False
+        return value > limit_value
+
+
+class DatasetConfig(BaseConfig):
+    """Manages the config attribute associated with a dataset."""
+
+    def __init__(self, tag, app_config, default_config):
+        super().__init__(app_config, default_config)
+        self.tag = tag
```

Comment:
the dataroot tag must be a legal URL path element.  I can't find code that validates that this is true, and it is probably worth adding.  Will help catch malformed configuration.

Reply:
The check was added in the original PR for multi data root.  It's in the checks for multi_dataset/dataroot attribute.  There is also a check that the tag in DatasetConfig must be one of the keys in multi_dataset/dataroot.   

I think given the previous discussion that plan is to add a  "URL_path" to the dataset config.  I'd then move the check for legal URL path elements to that config attribute, which would be clearer.